### PR TITLE
[WIP Issue 450] add trainable to module in order to support "freeze layers" in finetune

### PR DESCRIPTION
**Pull request by SeaOfOcean**
## What changes were proposed in this pull request?

add a trainable parameter to AbstractModule, default = true

When doing backward, if the module is not trainable, accGradParameters is ignored

When updating the parameters in optim, getParameters(filterTrainable=true) (filterTrainable is false by default) is called, and only trainable parameters are collected and updated.

## How was this patch tested?

unit test

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/450

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 1]
```diff
@@ -236,6 +287,16 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
 
   /**
+   * This function returns the array of trainable states. each element corresponds to
+   * the element of weights parameters
+   * If module only has weight or only has bias, then the trainalbles.length == 1
+   * If module contains both weight and bias, then the trainalbles.length == 2, and they should
+   * be both true or false
+   * @return trainables
+   */
+  def trainables(): Array[Boolean] = null
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: why not use something like `if (trainables) parameters()._1 else null`

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala [hunk 2]
```diff
@@ -658,6 +658,14 @@ class BatchNormalization[@specialized(Float, Double) T: ClassTag](
     }
   }
 
+  override def trainables(): Array[Boolean] = {
+    if (affine) {
+      Array(trainable, trainable)
+    } else {
+      null
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala]: Is empty array better here?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 3]
```diff
@@ -130,6 +131,26 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
     }
   }
 
+  protected var trainable = true
+
+  /**
+   * set trainable
+   * if trainable is true, train this module
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: please explain what do you mean of train this module?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Concat.scala [hunk 4]
```diff
@@ -140,10 +140,12 @@ class Concat[T: ClassTag](val dimension: Int)(
     var i = 0
     while (i < this.modules.length) {
       val currentOutput = this.modules(i).output.asInstanceOf[Tensor[T]]
-      this.modules(i).accGradParameters(
-        input.asInstanceOf[Activity],
-        gradOutput.narrow(dimension, offset, currentOutput.size(dimension))
-          .asInstanceOf[Activity], scale)
+      if (this.modules(i).isTrainable()) {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Concat.scala]: also change the behavior in backward()?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala (hunk 3 above)
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: do normal backward in the training process

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialConvolution.scala [hunk 5]
```diff
@@ -90,6 +90,8 @@ class SpatialConvolution[T: ClassTag](
   @transient
   protected var results: Array[Future[Unit]] = null
 
+  setTrainable(propagateBack)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialConvolution.scala]: do we still need propagateBack?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TimeDistributed.scala [hunk 6]
```diff
@@ -162,7 +162,8 @@ class TimeDistributed[T : ClassTag] (layer: TensorModule[T])
    *
    * @return
    */
-  override def getParameters(): (Tensor[T], Tensor[T]) = layer.getParameters()
+  override def getParameters(filterTrainable: Boolean = false)
+  : (Tensor[T], Tensor[T]) = layer.getParameters()
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TimeDistributed.scala]: why not layer.getParameters(filterTrainable)?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala (hunk 2 above)
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala]: Is empty array better here?

Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala]: Is empty array better here?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala (hunk 3 above)
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: what do you mean normal???

Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: Do you such general words in the docs.

Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: You should update the doc to make it more clear.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialConvolution.scala (hunk 5 above)
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialConvolution.scala]: in SpatialConvolution, if propagateBack = false, the gradInput is not computed.

if trainable = false, accGradParameters is not called.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TimeDistributed.scala (hunk 6 above)
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TimeDistributed.scala]: should have

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Concat.scala (hunk 4 above)
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Concat.scala]: done

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 7]
```diff
@@ -236,6 +263,18 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
   def parameters(): (Array[Tensor[T]], Array[Tensor[T]]) = null
 
   /**
+   * This function returns the array of trainable parameters.
+   * @return trainable parameters
+   */
+  def trainableParameters(): (Array[Tensor[T]], Array[Tensor[T]]) = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: why not add an optional flag to the parameter API?

**Comment by yiheng** [comment]
what if some people write a new container/module but doesn't use trainable variable to control its behavior? any way to give a compile/runtime error?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala (hunk 7 above)
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: parameters() is override in those layers with weight or bias, if I use optional flag, I need to modify many files in order to filter trainable. Instead, trainableParameters is only needed in AbstractModule.scala and Container.scala

**Comment by SeaOfOcean** [comment]
only newly added container needs to control trainable behavior

**Comment by yiheng** [comment]
This should be duplicated with PR #969
